src/identity/ldap/ldap.strategy.ts

```diff
@@ -0,0 +1,226 @@
+          searchFilter: ldapConfig.searchFilter,
+          searchAttributes: ldapConfig.searchAttributes,
+          bindDN: ldapConfig.bindDn,
+          bindCredentials: ldapConfig.bindCredentials,
+          tlsOptions: {
+            ca: ldapConfig.tlsCaCerts,
+          },
+        });
+        const username = req.body.username;
+        const password = req.body.password;
+        auth.once('error', (error) => {
+          throw new InternalServerErrorException(error);
+        });
+        // eslint-disable-next-line @typescript-eslint/no-empty-function
+        auth.on('error', () => {}); // Ignore further errors
+        auth.authenticate(
+          username,
+          password,
+          (error, user: Record<string, string>) => {
+            auth.close(() => {
+              // We don't care about the closing
+            });
+            if (error) {
+              this.handleLDAPError(username, error);
+            }
+
+            if (!user) {
+              throw new UnauthorizedException(LDAP_ERROR_MAP['default']);
+            }
+
+            const userId = user[ldapConfig.userIdField];
+
+            identityService
+              .getIdentityFromUserIdAndProviderType(userId, ProviderType.LDAP)
+              .then(async (identity) => {
+                await this.syncIdentity(
+                  identity,
+                  ldapConfig.displayNameField,
+                  ldapConfig.profilePictureField,
+                  user,
+                );
+                done(null, username);
+              })
+              .catch(async (e) => {
+                if (e instanceof NotInDBError) {
+                  // The user/identity does not yet exist
+                  const newUser = await this.usersService.createUser(
+                    username,
+                    // if there is no displayName we use the username
+                    user[ldapConfig.displayNameField] ?? username,
+                  );
+                  const identity = await identityService.createIdentity(
+                    newUser,
+                    ProviderType.LDAP,
+                    userId,
+                  );
+                  await this.syncIdentity(
+                    identity,
+                    ldapConfig.displayNameField,
+                    ldapConfig.profilePictureField,
+                    user,
+                  );
+                  done(null, username);
+                } else {
+                  throw e;
+                }
+              });
+          },
+        );
+      },
+    );
+  }
+
+  /**
+   * @async
+   * Sync identity with data from the ldap user.
+   * @param {Identity} identity - the identity to sync
+   * @param {string} displayNameField - the field to be used as a display name
+   * @param {string} profilePictureField - the field to be used as a profile picture
+   * @param {Record<string, string>} user - the user object from ldap
+   * @private
+   */
+  private async syncIdentity(
```

Comment:
As stated above `sync` might be confusing.

Reply:
changed